// @farcaster/auth-kit/dist/auth-kit.js
import { viemConnector as Se, createAppClient as be } from "@farcaster/auth-client";
export * from "@farcaster/auth-client";
import { jsx as C, jsxs as T, Fragment as J } from "react/jsx-runtime";
import * as I from "react";
import { createContext as Be, useState as P, useEffect as Z, useCallback as R, useContext as Ie, useRef as Jt, useMemo as Ae } from "react";
import { createPortal as Me } from "react-dom";
const Ne = typeof window < "u" && (window != null && window.location) ? {
  domain: window.location.host,
  siweUri: window.location.href
} : {}, Wt = {
  relay: "https://relay.farcaster.xyz",
  version: "v1",
  ...Ne
}, qt = Be({
  isAuthenticated: !1,
  config: Wt,
  profile: {},
  signInMessage: {},
  onSignIn: () => {
  },
  onSignOut: () => {
  }
});
function Pe({
  config: t,
  children: e
}) {
  const [n, r] = P(), [o, i] = P(!1), [a, c] = P({}), [s, f] = P({}), l = {
    ...Wt,
    ...t
  }, { relay: E, rpcUrl: w, rpcUrls: d, version: g } = l;
  Z(() => {
    const h = Se({ rpcUrl: w, rpcUrls: d }), p = be({
      relay: E,
      ethereum: h,
      version: g
    });
    r(p);
  }, [E, w, d, g]);
  const u = R((h) => {
    const { message: p, signature: v, fid: m, username: S, bio: B, displayName: b, pfpUrl: A, custody: M, verifications: N } = h;
    i(!0), c({ fid: m, username: S, bio: B, displayName: b, pfpUrl: A, custody: M, verifications: N }), f({ message: p, signature: v });
  }, []), y = () => {
    i(!1), c({}), f({});
  };
  return /* @__PURE__ */ C(
    qt.Provider,
    {
      value: {
        appClient: n,
        isAuthenticated: o,
        profile: a,
        signInMessage: s,
        config: l,
        onSignIn: u,
        onSignOut: y
      },
      children: e
    }
  );
}
function q() {
  return Ie(qt);
}
function lt() {
  const { appClient: t } = q();
  return t;
}
function Te({
  nonce: t,
  notBefore: e,
  expirationTime: n,
  requestId: r,
  onSuccess: o,
  onError: i,
  redirectUrl: a
}) {
  const { config: c } = q(), { siweUri: s, domain: f } = c, l = lt(), [E, w] = P(!1), [d, g] = P(), [u, y] = P(), [h, p] = P(), [v, m] = P(!1), [S, B] = P(!1), [b, A] = P(), M = R(async () => {
    if (E && l && s && f && !d) {
      const _ = typeof t == "function" ? await t() : t, {
        data: mt,
        isError: vt,
        error: X
      } = await l.createChannel({
        nonce: _,
        siweUri: s,
        domain: f,
        notBefore: e,
        expirationTime: n,
        requestId: r,
        redirectUrl: a
      });
      if (vt)
        B(!0), A(X), i == null || i(X);
      else {
        const { channelToken: tt, url: Zt, nonce: Dt } = mt;
        g(tt), y(Zt), p(Dt), m(!0), o == null || o({ channelToken: tt, url: Zt, nonce: Dt });
      }
    }
  }, [
    E,
    l,
    s,
    f,
    d,
    t,
    e,
    n,
    r,
    i,
    o,
    a
  ]);
  Z(() => {
    M();
  }, [M]);
  const N = R(async () => {
    w(!0);
  }, [w]), k = R(() => {
    g(void 0), y(void 0), m(!1), B(!1), A(void 0);
  }, [g, y, m, B, A]), L = R(() => {
    k(), N();
  }, [N, k]);
  return {
    connect: N,
    reconnect: L,
    reset: k,
    isSuccess: v,
    isError: S,
    error: b,
    data: { channelToken: d, url: u, nonce: h }
  };
}
function Re({
  nonce: t,
  domain: e,
  message: n,
  signature: r,
  onSuccess: o,
  onError: i
}) {
  const a = lt(), [c, s] = P(!1), [f, l] = P(!1), [E, w] = P(!1), [d, g] = P(), u = async () => {
    w(!1), l(!1), s(!1), g(void 0);
  }, y = R(async () => {
    if (a && t && e && n && r) {
      const {
        success: h,
        isError: p,
        error: v
      } = await a.verifySignInMessage({
        nonce: t,
        domain: e,
        message: n,
        signature: r
      });
      p ? (w(!0), g(v), i == null || i(v)) : (s(h), l(!0), o == null || o({ message: n, signature: r, validSignature: h }));
    }
  }, [a, t, e, n, r, o, i]);
  return Z(() => {
    u(), t && e && n && r && y();
  }, [t, e, n, r, y]), {
    isSuccess: f,
    isError: E,
    error: d,
    data: { message: n, signature: r, validSignature: c }
  };
}
const ke = {
  timeout: 3e5,
  interval: 1500
};
function Le(t) {
  const e = lt(), { channelToken: n, timeout: r, interval: o, onSuccess: i, onError: a, onResponse: c } = {
    ...ke,
    ...t
  }, [s, f] = P(), [l, E] = P(!1), [w, d] = P(!1), [g, u] = P(!1), [y, h] = P(!1), [p, v] = P(), m = () => {
    f(void 0), E(!1), h(!1), d(!1), v(void 0);
  }, S = () => {
    E(!0);
  }, B = R(async () => {
    if (l && e && n) {
      u(!0);
      const {
        data: b,
        isError: A,
        error: M
      } = await e.watchStatus({
        channelToken: n,
        timeout: r,
        interval: o,
        onResponse: ({ data: N }) => {
          f(N), c == null || c(N);
        }
      });
      A ? (h(!0), u(!1), v(M), a == null || a(M)) : (d(!0), u(!1), f(b), i == null || i(b));
    }
  }, [l, e, n, r, o, i, a, c]);
  return Z(() => {
    m(), n && B();
  }, [n, B]), {
    watch: S,
    isSuccess: w,
    isPolling: g,
    isError: y,
    error: p,
    data: s
  };
}
const _e = {
  timeout: 3e5,
  interval: 1500
};
function Fe(t) {
  const e = lt(), {
    onSignIn: n,
    onSignOut: r,
    config: { domain: o, redirectUrl: i }
  } = q(), { timeout: a, interval: c, onSuccess: s, onStatusResponse: f, onError: l, ...E } = {
    ..._e,
    ...t
  }, {
    connect: w,
    reconnect: d,
    reset: g,
    data: { channelToken: u, url: y, nonce: h },
    isSuccess: p,
    isError: v,
    error: m
  } = Te({ ...E, redirectUrl: i, onError: l }), {
    watch: S,
    isPolling: B,
    data: b,
    isError: A,
    error: M
  } = Le({
    channelToken: u,
    timeout: a,
    interval: c,
    onError: l,
    onResponse: f
  }), {
    isSuccess: N,
    data: { validSignature: k },
    isError: L,
    error: _
  } = Re({
    nonce: h,
    domain: o,
    message: b == null ? void 0 : b.message,
    signature: b == null ? void 0 : b.signature,
    onError: l
  }), mt = v || A || L, vt = m || M || _, X = R(() => {
    S();
  }, [S]), tt = R(() => {
    r(), g();
  }, [r, g]);
  return Z(() => {
    N && b && k && (n(b), s == null || s(b));
  }, [N, b, k, n, s]), {
    signIn: X,
    signOut: tt,
    connect: w,
    reconnect: d,
    isConnected: p,
    isSuccess: N,
    isPolling: B,
    isError: mt,
    error: vt,
    channelToken: u,
    url: y,
    appClient: e,
    data: b,
    validSignature: k
  };
}
function Ze({
  height: t,
  fill: e
}) {
  const n = t * 1.1;
  return /* @__PURE__ */ T(
    "svg",
    {
      xmlns: "http://www.w3.org/2000/svg",
      width: n,
      height: t,
      viewBox: "0 0 22 20",
      fill: "none",
      children: [
        /* @__PURE__ */ C("title", { children: "Farcaster logo" }),
        /* @__PURE__ */ T("g", { fill: Ut[e], clipPath: "url(#a)", children: [
          /* @__PURE__ */ C("path", { d: "M3.786.05h14.156v2.824h4.025l-.844 2.825h-.714v11.427c.358 0 .65.287.65.642v.77h.13c.358 0 .649.288.649.642v.77h-7.273v-.77c0-.354.29-.642.65-.642h.13v-.77c0-.309.22-.566.512-.628l-.014-6.306c-.23-2.519-2.37-4.493-4.98-4.493-2.608 0-4.75 1.974-4.979 4.494l-.013 6.3c.346.05.772.315.772.633v.77h.13c.358 0 .65.288.65.642v.77H.15v-.77c0-.354.29-.642.649-.642h.13v-.77c0-.355.29-.642.65-.642V5.7H.863L.02 2.874h3.766V.05Z" }),
          /* @__PURE__ */ C("path", { d: "M17.942.05h.047V.003h-.047V.05ZM3.786.05V.003H3.74V.05h.047Zm14.156 2.824h-.048v.047h.048v-.047Zm4.025 0 .046.013.018-.06h-.064v.047Zm-.844 2.824v.047h.035l.01-.033-.045-.014Zm-.714 0v-.046h-.047v.046h.047Zm0 11.428h-.047v.047h.047v-.047Zm.65 1.412h-.048v.047h.047v-.047Zm.779 1.412v.047h.046v-.047h-.046Zm-7.273 0h-.048v.047h.048v-.047Zm.78-1.412v.047h.047v-.047h-.048Zm.512-1.398.01.045.038-.007v-.038h-.048Zm-.014-6.306h.048v-.004l-.048.004Zm-9.959 0-.047-.004v.004h.047Zm-.013 6.3h-.048v.041l.04.006.008-.047Zm.772 1.404h-.047v.047h.047v-.047Zm.78 1.412v.047h.047v-.047h-.048Zm-7.273 0H.102v.047H.15v-.047Zm.779-1.412v.047h.047v-.047H.93Zm.65-1.412v.047h.047v-.047h-.047Zm0-11.428h.047v-.046h-.047v.046Zm-.715 0-.045.014.01.033h.035v-.047ZM.02 2.874v-.047h-.063l.018.06.045-.013Zm3.766 0v.047h.048v-.047h-.048ZM17.942.003H3.786v.093h14.156V.003Zm.047 2.87V.05h-.095v2.824h.095Zm3.978-.046h-4.025v.094h4.025v-.094Zm-.798 2.885.844-2.825-.091-.026-.845 2.824.092.027Zm-.76.033h.714v-.093h-.714v.093Zm.048 11.381V5.698h-.095v11.428h.095Zm.649.641a.693.693 0 0 0-.697-.688v.094c.332 0 .602.266.602.595h.095Zm0 .77v-.77h-.095v.77h.095Zm.082-.045h-.13v.093h.13v-.093Zm.697.688a.692.692 0 0 0-.697-.688v.093c.333 0 .602.267.602.595h.095Zm0 .77v-.77h-.095v.77h.095Zm-3.943.047h3.896v-.094h-3.896v.094Zm-2.079 0h2.079v-.094h-2.079v.094Zm-1.298 0h1.298v-.094h-1.298v.094Zm-.048-.817v.77h.095v-.77h-.095Zm.697-.688a.693.693 0 0 0-.697.688h.095c0-.328.27-.595.602-.595v-.093Zm.13 0h-.13v.093h.13v-.093Zm-.047-.725v.77h.095v-.77h-.095Zm.55-.673a.69.69 0 0 0-.55.673h.095c0-.285.203-.524.476-.582l-.02-.09Zm-.051-6.26.014 6.306h.095l-.014-6.306h-.095Zm-4.932-4.447c2.583 0 4.704 1.956 4.932 4.452l.094-.009c-.232-2.543-2.393-4.536-5.026-4.536v.093ZM5.932 10.84c.227-2.496 2.348-4.452 4.932-4.452v-.093c-2.633 0-4.795 1.993-5.027 4.536l.095.008Zm-.014 6.295.014-6.3h-.095l-.014 6.3h.095Zm.773.633c0-.18-.12-.337-.276-.453a1.236 1.236 0 0 0-.538-.226l-.014.093c.166.025.351.1.495.207.145.108.238.241.238.38h.095Zm0 .77v-.77h-.095v.77h.095Zm.082-.045h-.13v.093h.13v-.093Zm.697.688a.693.693 0 0 0-.697-.688v.093c.332 0 .602.267.602.595h.095Zm0 .77v-.77h-.095v.77h.095Zm-1.606.047h1.558v-.094H5.864v.094Zm-.108 0h.108v-.094h-.108v.094Zm-1.97 0h1.97v-.094h-1.97v.094Zm-3.636 0h3.636v-.094H.15v.094Zm-.048-.817v.77h.095v-.77H.102Zm.697-.688a.693.693 0 0 0-.697.688h.095c0-.328.27-.595.602-.595v-.093Zm.13 0h-.13v.093h.13v-.093Zm-.048-.725v.77h.095v-.77H.881Zm.698-.688a.693.693 0 0 0-.698.688h.095c0-.328.27-.594.603-.594v-.094ZM1.53 5.7v11.427h.095V5.698h-.095Zm-.667.046h.715v-.093H.864v.093Zm-.89-2.858L.82 5.712l.09-.027-.844-2.824-.09.026Zm3.812-.06H.02v.094h3.766v-.094ZM3.74.05v2.824h.095V.05h-.095Z" })
        ] }),
        /* @__PURE__ */ C("defs", { children: /* @__PURE__ */ C("clipPath", { id: "a", children: /* @__PURE__ */ C("path", { fill: Ut[e], d: "M0 0h22v20H0z" }) }) })
      ]
    }
  );
}
const Ut = {
  purple: "#7C65C1",
  white: "#FFFFFF"
};
var De = "_1n3pr301", Ue = "_1n3pr305", Ve = "_1n3pr304", He = "_1n3pr300", Tt = "_1n3pr302", Oe = "_1n3pr303";
function Qt({ kind: t = "primary", children: e, className: n, ...r }) {
  return /* @__PURE__ */ C("button", { type: "button", className: xe[t] || n, ...r, children: e });
}
const xe = {
  primary: De,
  secondary: Tt,
  tertiary: Oe,
  reset: He
};
function ze({
  label: t,
  onClick: e,
  initializing: n
}) {
  return /* @__PURE__ */ T(Qt, { kind: "primary", onClick: e, disabled: n, children: [
    /* @__PURE__ */ C(Ze, { height: 20, fill: "white" }),
    /* @__PURE__ */ C("span", { style: { marginLeft: 9 }, children: t })
  ] });
}
function Ke(t, e) {
  Z(() => {
    function n(r) {
      t.current && !t.current.contains(r.target) && e();
    }
    return document.addEventListener("mousedown", n), () => {
      document.removeEventListener("mousedown", n);
    };
  }, [t, e]);
}
var Ye = "_1rbg3ua1", $e = "_1rbg3ua2", je = "_1rbg3ua3", Je = "_16wbil70", We = "_16wbil71";
function qe({ signOut: t }) {
  return /* @__PURE__ */ C("div", { className: Je, children: /* @__PURE__ */ T(
    "button",
    {
      type: "button",
      className: Tt,
      style: {
        boxShadow: "0px 6px 12px 0 rgba(0,0,0,0.12)"
      },
      onClick: t,
      children: [
        /* @__PURE__ */ T(
          "svg",
          {
            width: "19",
            height: "20",
            viewBox: "0 0 19 20",
            fill: "none",
            xmlns: "http://www.w3.org/2000/svg",
            className: We,
            children: [
              /* @__PURE__ */ C("title", { children: "Sign Out" }),
              /* @__PURE__ */ C(
                "path",
                {
                  d: "M11.75 7V3.25C11.75 2.65326 11.5129 2.08097 11.091 1.65901C10.669 1.23705 10.0967 1 9.5 1H3.5C2.90326 1 2.33097 1.23705 1.90901 1.65901C1.48705 2.08097 1.25 2.65326 1.25 3.25V16.75C1.25 17.3467 1.48705 17.919 1.90901 18.341C2.33097 18.7629 2.90326 19 3.5 19H9.5C10.0967 19 10.669 18.7629 11.091 18.341C11.5129 17.919 11.75 17.3467 11.75 16.75V13M14.75 13L17.75 10M17.75 10L14.75 7M17.75 10H5",
                  stroke: "black",
                  strokeWidth: "1.5",
                  strokeLinecap: "round",
                  strokeLinejoin: "round"
                }
              )
            ]
          }
        ),
        "Sign out"
      ]
    }
  ) });
}
function Qe({
  userData: t,
  signOut: e,
  hideSignOut: n
}) {
  const [r, o] = P(!1), i = Jt(null);
  Ke(i, () => o(!1));
  const a = (t == null ? void 0 : t.username) ?? `!${t == null ? void 0 : t.fid}`, c = (t == null ? void 0 : t.pfpUrl) ?? "https://farcaster.xyz/avatar.png", s = r && !n;
  return /* @__PURE__ */ C(
    "div",
    {
      className: `fc-authkit-profile-button ${Ye}`,
      ref: i,
      children: /* @__PURE__ */ T(
        "div",
        {
          style: {
            display: "flex",
            flexDirection: "column",
            alignItems: "flex-end"
          },
          children: [
            /* @__PURE__ */ T(
              "button",
              {
                type: "button",
                className: Tt,
                onClick: () => o(!r),
                children: [
                  /* @__PURE__ */ C("img", { className: $e, src: c, alt: "avatar" }),
                  /* @__PURE__ */ C("span", { className: je, children: a }),
                  !n && /* @__PURE__ */ T(
                    "svg",
                    {
                      width: "24",
                      height: "24",
                      viewBox: "0 0 24 24",
                      fill: "none",
                      xmlns: "http://www.w3.org/2000/svg",
                      children: [
                        /* @__PURE__ */ C("title", { children: "Dropdown" }),
                        /* @__PURE__ */ C("g", { opacity: "0.5", children: /* @__PURE__ */ C(
                          "path",
                          {
                            fillRule: "evenodd",
                            clipRule: "evenodd",
                            d: "M6.26418 9.864C6.43293 9.69545 6.66168 9.60079 6.90018 9.60079C7.13869 9.60079 7.36743 9.69545 7.53618 9.864L12.0002 14.328L16.4642 9.864C16.5466 9.77557 16.6459 9.70465 16.7563 9.65546C16.8667 9.60627 16.9859 9.57982 17.1068 9.57769C17.2276 9.57555 17.3476 9.59778 17.4597 9.64305C17.5718 9.68831 17.6736 9.75569 17.759 9.84115C17.8445 9.92661 17.9119 10.0284 17.9571 10.1405C18.0024 10.2525 18.0246 10.3726 18.0225 10.4934C18.0204 10.6143 17.9939 10.7334 17.9447 10.8438C17.8955 10.9542 17.8246 11.0536 17.7362 11.136L12.6362 16.236C12.4674 16.4045 12.2387 16.4992 12.0002 16.4992C11.7617 16.4992 11.5329 16.4045 11.3642 16.236L6.26418 11.136C6.09564 10.9672 6.00098 10.7385 6.00098 10.5C6.00098 10.2615 6.09564 10.0327 6.26418 9.864Z",
                            fill: "black"
                          }
                        ) })
                      ]
                    }
                  )
                ]
              }
            ),
            s && /* @__PURE__ */ C(qe, { signOut: e })
          ]
        }
      )
    }
  );
}
var D = function() {
  return D = Object.assign || function(e) {
    for (var n, r = 1, o = arguments.length; r < o; r++) {
      n = arguments[r];
      for (var i in n) Object.prototype.hasOwnProperty.call(n, i) && (e[i] = n[i]);
    }
    return e;
  }, D.apply(this, arguments);
};
function Gt(t, e) {
  var n = {};
  for (var r in t) Object.prototype.hasOwnProperty.call(t, r) && e.indexOf(r) < 0 && (n[r] = t[r]);
  if (t != null && typeof Object.getOwnPropertySymbols == "function")
    for (var o = 0, r = Object.getOwnPropertySymbols(t); o < r.length; o++)
      e.indexOf(r[o]) < 0 && Object.prototype.propertyIsEnumerable.call(t, r[o]) && (n[r[o]] = t[r[o]]);
  return n;
}
function Ge(t, e, n) {
  if (n || arguments.length === 2) for (var r = 0, o = e.length, i; r < o; r++)
    (i || !(r in e)) && (i || (i = Array.prototype.slice.call(e, 0, r)), i[r] = e[r]);
  return t.concat(i || Array.prototype.slice.call(e));
}
var it = "right-scroll-bar-position", at = "width-before-scroll-bar", Xe = "with-scroll-bars-hidden", tn = "--removed-body-scroll-bar-size";
function en(t, e) {
  return typeof t == "function" ? t(e) : t && (t.current = e), t;
}
function nn(t, e) {
  var n = P(function() {
    return {
      // value
      value: t,
      // last callback
      callback: e,
      // "memoized" public interface
      facade: {
        get current() {
          return n.value;
        },
        set current(r) {
          var o = n.value;
          o !== r && (n.value = r, n.callback(r, o));
        }
      }
    };
  })[0];
  return n.callback = e, n.facade;
}
function rn(t, e) {
  return nn(null, function(n) {
    return t.forEach(function(r) {
      return en(r, n);
    });
  });
}
function on(t) {
  return t;
}
function an(t, e) {
  e === void 0 && (e = on);
  var n = [], r = !1, o = {
    read: function() {
      if (r)
        throw new Error("Sidecar: could not `read` from an `assigned` medium. `read` could be used only with `useMedium`.");
      return n.length ? n[n.length - 1] : t;
    },
    useMedium: function(i) {
      var a = e(i, r);
      return n.push(a), function() {
        n = n.filter(function(c) {
          return c !== a;
        });
      };
    },
    assignSyncMedium: function(i) {
      for (r = !0; n.length; ) {
        var a = n;
        n = [], a.forEach(i);
      }
      n = {
        push: function(c) {
          return i(c);
        },
        filter: function() {
          return n;
        }
      };
    },
    assignMedium: function(i) {
      r = !0;
      var a = [];
      if (n.length) {
        var c = n;
        n = [], c.forEach(i), a = n;
      }
      var s = function() {
        var l = a;
        a = [], l.forEach(i);
      }, f = function() {
        return Promise.resolve().then(s);
      };
      f(), n = {
        push: function(l) {
          a.push(l), f();
        },
        filter: function(l) {
          return a = a.filter(l), n;
        }
      };
    }
  };
  return o;
}
function cn(t) {
  t === void 0 && (t = {});
  var e = an(null);
  return e.options = D({ async: !0, ssr: !1 }, t), e;
}
var Xt = function(t) {
  var e = t.sideCar, n = Gt(t, ["sideCar"]);
  if (!e)
    throw new Error("Sidecar: please provide `sideCar` property to import the right car");
  var r = e.read();
  if (!r)
    throw new Error("Sidecar medium not found");
  return I.createElement(r, D({}, n));
};
Xt.isSideCarExport = !0;
function sn(t, e) {
  return t.useMedium(e), Xt;
}
var te = cn(), wt = function() {
}, ut = I.forwardRef(function(t, e) {
  var n = I.useRef(null), r = I.useState({
    onScrollCapture: wt,
    onWheelCapture: wt,
    onTouchMoveCapture: wt
  }), o = r[0], i = r[1], a = t.forwardProps, c = t.children, s = t.className, f = t.removeScrollBar, l = t.enabled, E = t.shards, w = t.sideCar, d = t.noIsolation, g = t.inert, u = t.allowPinchZoom, y = t.as, h = y === void 0 ? "div" : y, p = t.gapMode, v = Gt(t, ["forwardProps", "children", "className", "removeScrollBar", "enabled", "shards", "sideCar", "noIsolation", "inert", "allowPinchZoom", "as", "gapMode"]), m = w, S = rn([n, e]), B = D(D({}, v), o);
  return I.createElement(
    I.Fragment,
    null,
    l && I.createElement(m, { sideCar: te, removeScrollBar: f, shards: E, noIsolation: d, inert: g, setCallbacks: i, allowPinchZoom: !!u, lockRef: n, gapMode: p }),
    a ? I.cloneElement(I.Children.only(c), D(D({}, B), { ref: S })) : I.createElement(h, D({}, B, { className: s, ref: S }), c)
  );
});
ut.defaultProps = {
  enabled: !0,
  removeScrollBar: !0,
  inert: !1
};
ut.classNames = {
  fullWidth: at,
  zeroRight: it
};
var ln = function() {
  if (typeof __webpack_nonce__ < "u")
    return __webpack_nonce__;
};
function un() {
  if (!document)
    return null;
  var t = document.createElement("style");
  t.type = "text/css";
  var e = ln();
  return e && t.setAttribute("nonce", e), t;
}
function fn(t, e) {
  t.styleSheet ? t.styleSheet.cssText = e : t.appendChild(document.createTextNode(e));
}
function dn(t) {
  var e = document.head || document.getElementsByTagName("head")[0];
  e.appendChild(t);
}
var hn = function() {
  var t = 0, e = null;
  return {
    add: function(n) {
      t == 0 && (e = un()) && (fn(e, n), dn(e)), t++;
    },
    remove: function() {
      t--, !t && e && (e.parentNode && e.parentNode.removeChild(e), e = null);
    }
  };
}, gn = function() {
  var t = hn();
  return function(e, n) {
    I.useEffect(function() {
      return t.add(e), function() {
        t.remove();
      };
    }, [e && n]);
  };
}, ee = function() {
  var t = gn(), e = function(n) {
    var r = n.styles, o = n.dynamic;
    return t(r, o), null;
  };
  return e;
}, mn = {
  left: 0,
  top: 0,
  right: 0,
  gap: 0
}, pt = function(t) {
  return parseInt(t || "", 10) || 0;
}, vn = function(t) {
  var e = window.getComputedStyle(document.body), n = e[t === "padding" ? "paddingLeft" : "marginLeft"], r = e[t === "padding" ? "paddingTop" : "marginTop"], o = e[t === "padding" ? "paddingRight" : "marginRight"];
  return [pt(n), pt(r), pt(o)];
}, wn = function(t) {
  if (t === void 0 && (t = "margin"), typeof window > "u")
    return mn;
  var e = vn(t), n = document.documentElement.clientWidth, r = window.innerWidth;
  return {
    left: e[0],
    top: e[1],
    right: e[2],
    gap: Math.max(0, r - n + e[2] - e[0])
  };
}, pn = ee(), yn = function(t, e, n, r) {
  var o = t.left, i = t.top, a = t.right, c = t.gap;
  return n === void 0 && (n = "margin"), `
  .`.concat(Xe, ` {
   overflow: hidden `).concat(r, `;
   padding-right: `).concat(c, "px ").concat(r, `;
  }
  body {
    overflow: hidden `).concat(r, `;
    overscroll-behavior: contain;
    `).concat([
    e && "position: relative ".concat(r, ";"),
    n === "margin" && `
    padding-left: `.concat(o, `px;
    padding-top: `).concat(i, `px;
    padding-right: `).concat(a, `px;
    margin-left:0;
    margin-top:0;
    margin-right: `).concat(c, "px ").concat(r, `;
    `),
    n === "padding" && "padding-right: ".concat(c, "px ").concat(r, ";")
  ].filter(Boolean).join(""), `
  }
  
  .`).concat(it, ` {
    right: `).concat(c, "px ").concat(r, `;
  }
  
  .`).concat(at, ` {
    margin-right: `).concat(c, "px ").concat(r, `;
  }
  
  .`).concat(it, " .").concat(it, ` {
    right: 0 `).concat(r, `;
  }
  
  .`).concat(at, " .").concat(at, ` {
    margin-right: 0 `).concat(r, `;
  }
  
  body {
    `).concat(tn, ": ").concat(c, `px;
  }
`);
}, Cn = function(t) {
  var e = t.noRelative, n = t.noImportant, r = t.gapMode, o = r === void 0 ? "margin" : r, i = I.useMemo(function() {
    return wn(o);
  }, [o]);
  return I.createElement(pn, { styles: yn(i, !e, o, n ? "" : "!important") });
}, Bt = !1;
if (typeof window < "u")
  try {
    var et = Object.defineProperty({}, "passive", {
      get: function() {
        return Bt = !0, !0;
      }
    });
    window.addEventListener("test", et, et), window.removeEventListener("test", et, et);
  } catch {
    Bt = !1;
  }
var O = Bt ? { passive: !1 } : !1, En = function(t) {
  return t.tagName === "TEXTAREA";
}, ne = function(t, e) {
  var n = window.getComputedStyle(t);
  return (
    // not-not-scrollable
    n[e] !== "hidden" && // contains scroll inside self
    !(n.overflowY === n.overflowX && !En(t) && n[e] === "visible")
  );
}, Sn = function(t) {
  return ne(t, "overflowY");
}, bn = function(t) {
  return ne(t, "overflowX");
}, Vt = function(t, e) {
  var n = e.ownerDocument, r = e;
  do {
    typeof ShadowRoot < "u" && r instanceof ShadowRoot && (r = r.host);
    var o = re(t, r);
    if (o) {
      var i = oe(t, r), a = i[1], c = i[2];
      if (a > c)
        return !0;
    }
    r = r.parentNode;
  } while (r && r !== n.body);
  return !1;
}, Bn = function(t) {
  var e = t.scrollTop, n = t.scrollHeight, r = t.clientHeight;
  return [
    e,
    n,
    r
  ];
}, In = function(t) {
  var e = t.scrollLeft, n = t.scrollWidth, r = t.clientWidth;
  return [
    e,
    n,
    r
  ];
}, re = function(t, e) {
  return t === "v" ? Sn(e) : bn(e);
}, oe = function(t, e) {
  return t === "v" ? Bn(e) : In(e);
}, An = function(t, e) {
  return t === "h" && e === "rtl" ? -1 : 1;
}, Mn = function(t, e, n, r, o) {
  var i = An(t, window.getComputedStyle(e).direction), a = i * r, c = n.target, s = e.contains(c), f = !1, l = a > 0, E = 0, w = 0;
  do {
    var d = oe(t, c), g = d[0], u = d[1], y = d[2], h = u - y - i * g;
    (g || h) && re(t, c) && (E += h, w += g), c instanceof ShadowRoot ? c = c.host : c = c.parentNode;
  } while (
    // portaled content
    !s && c !== document.body || // self content
    s && (e.contains(c) || e === c)
  );
  return (l && Math.abs(E) < 1 || !l && Math.abs(w) < 1) && (f = !0), f;
}, nt = function(t) {
  return "changedTouches" in t ? [t.changedTouches[0].clientX, t.changedTouches[0].clientY] : [0, 0];
}, Ht = function(t) {
  return [t.deltaX, t.deltaY];
}, Ot = function(t) {
  return t && "current" in t ? t.current : t;
}, Nn = function(t, e) {
  return t[0] === e[0] && t[1] === e[1];
}, Pn = function(t) {
  return `
  .block-interactivity-`.concat(t, ` {pointer-events: none;}
  .allow-interactivity-`).concat(t, ` {pointer-events: all;}
`);
}, Tn = 0, x = [];
function Rn(t) {
  var e = I.useRef([]), n = I.useRef([0, 0]), r = I.useRef(), o = I.useState(Tn++)[0], i = I.useState(ee)[0], a = I.useRef(t);
  I.useEffect(function() {
    a.current = t;
  }, [t]), I.useEffect(function() {
    if (t.inert) {
      document.body.classList.add("block-interactivity-".concat(o));
      var u = Ge([t.lockRef.current], (t.shards || []).map(Ot), !0).filter(Boolean);
      return u.forEach(function(y) {
        return y.classList.add("allow-interactivity-".concat(o));
      }), function() {
        document.body.classList.remove("block-interactivity-".concat(o)), u.forEach(function(y) {
          return y.classList.remove("allow-interactivity-".concat(o));
        });
      };
    }
  }, [t.inert, t.lockRef.current, t.shards]);
  var c = I.useCallback(function(u, y) {
    if ("touches" in u && u.touches.length === 2)
      return !a.current.allowPinchZoom;
    var h = nt(u), p = n.current, v = "deltaX" in u ? u.deltaX : p[0] - h[0], m = "deltaY" in u ? u.deltaY : p[1] - h[1], S, B = u.target, b = Math.abs(v) > Math.abs(m) ? "h" : "v";
    if ("touches" in u && b === "h" && B.type === "range")
      return !1;
    var A = Vt(b, B);
    if (!A)
      return !0;
    if (A ? S = b : (S = b === "v" ? "h" : "v", A = Vt(b, B)), !A)
      return !1;
    if (!r.current && "changedTouches" in u && (v || m) && (r.current = S), !S)
      return !0;
    var M = r.current || S;
    return Mn(M, y, u, M === "h" ? v : m);
  }, []), s = I.useCallback(function(u) {
    var y = u;
    if (!(!x.length || x[x.length - 1] !== i)) {
      var h = "deltaY" in y ? Ht(y) : nt(y), p = e.current.filter(function(S) {
        return S.name === y.type && (S.target === y.target || y.target === S.shadowParent) && Nn(S.delta, h);
      })[0];
      if (p && p.should) {
        y.cancelable && y.preventDefault();
        return;
      }
      if (!p) {
        var v = (a.current.shards || []).map(Ot).filter(Boolean).filter(function(S) {
          return S.contains(y.target);
        }), m = v.length > 0 ? c(y, v[0]) : !a.current.noIsolation;
        m && y.cancelable && y.preventDefault();
      }
    }
  }, []), f = I.useCallback(function(u, y, h, p) {
    var v = { name: u, delta: y, target: h, should: p, shadowParent: kn(h) };
    e.current.push(v), setTimeout(function() {
      e.current = e.current.filter(function(m) {
        return m !== v;
      });
    }, 1);
  }, []), l = I.useCallback(function(u) {
    n.current = nt(u), r.current = void 0;
  }, []), E = I.useCallback(function(u) {
    f(u.type, Ht(u), u.target, c(u, t.lockRef.current));
  }, []), w = I.useCallback(function(u) {
    f(u.type, nt(u), u.target, c(u, t.lockRef.current));
  }, []);
  I.useEffect(function() {
    return x.push(i), t.setCallbacks({
      onScrollCapture: E,
      onWheelCapture: E,
      onTouchMoveCapture: w
    }), document.addEventListener("wheel", s, O), document.addEventListener("touchmove", s, O), document.addEventListener("touchstart", l, O), function() {
      x = x.filter(function(u) {
        return u !== i;
      }), document.removeEventListener("wheel", s, O), document.removeEventListener("touchmove", s, O), document.removeEventListener("touchstart", l, O);
    };
  }, []);
  var d = t.removeScrollBar, g = t.inert;
  return I.createElement(
    I.Fragment,
    null,
    g ? I.createElement(i, { styles: Pn(o) }) : null,
    d ? I.createElement(Cn, { gapMode: t.gapMode }) : null
  );
}
function kn(t) {
  for (var e = null; t !== null; )
    t instanceof ShadowRoot && (e = t.host, t = t.host), t = t.parentNode;
  return e;
}
const Ln = sn(te, Rn);
var ie = I.forwardRef(function(t, e) {
  return I.createElement(ut, D({}, t, { ref: e, sideCar: Ln }));
});
ie.classNames = ut.classNames;
var _n = "nao3n43", Fn = "nao3n42";
const xt = (t, e) => {
  const n = t.querySelectorAll("button:not(:disabled), a[href]");
  n.length !== 0 && n[e === "end" ? n.length - 1 : 0].focus();
};
function Zn(t) {
  const e = Jt(null);
  return Z(() => {
    const n = document.activeElement;
    return () => {
      var r;
      (r = n.focus) == null || r.call(n);
    };
  }, []), Z(() => {
    if (e.current) {
      const n = e.current.querySelector("[data-auto-focus]");
      n ? n.focus() : e.current.focus();
    }
  }, []), /* @__PURE__ */ T(J, { children: [
    /* @__PURE__ */ C(
      "div",
      {
        onFocus: R(() => e.current && xt(e.current, "end"), []),
        tabIndex: 0
      }
    ),
    /* @__PURE__ */ C("div", { ref: e, style: { outline: "none" }, tabIndex: -1, ...t }),
    /* @__PURE__ */ C(
      "div",
      {
        onFocus: R(() => e.current && xt(e.current, "start"), []),
        tabIndex: 0
      }
    )
  ] });
}
function Dn() {
  return typeof navigator < "u" && /android/i.test(navigator.userAgent);
}
function Un() {
  return typeof navigator < "u" && /iPhone|iPod/.test(navigator.userAgent);
}
function Vn() {
  return typeof navigator < "u" && (/iPad/.test(navigator.userAgent) || navigator.platform === "MacIntel" && navigator.maxTouchPoints > 1);
}
function Hn() {
  return Un() || Vn();
}
function It() {
  return Dn() || Hn();
}
const On = (t) => t.stopPropagation();
function xn({ children: t, onClose: e, open: n, titleId: r }) {
  Z(() => {
    const c = (s) => n && s.key === "Escape" && e();
    return document.addEventListener("keydown", c), () => document.removeEventListener("keydown", c);
  }, [n, e]);
  const [o, i] = P(!0);
  Z(() => {
    i(
      getComputedStyle(window.document.body).overflow !== "hidden"
    );
  }, []);
  const a = R(() => e(), [e]);
  return /* @__PURE__ */ C(J, { children: n ? Me(
    /* @__PURE__ */ C(ie, { enabled: o, children: /* @__PURE__ */ C(
      "div",
      {
        style: {
          alignItems: It() ? "flex-end" : "center",
          position: "fixed"
        },
        "aria-labelledby": r,
        "aria-modal": !0,
        className: Fn,
        onClick: a,
        role: "dialog",
        children: /* @__PURE__ */ C(
          Zn,
          {
            className: _n,
            onClick: On,
            role: "document",
            children: t
          }
        )
      }
    ) }),
    document.body
  ) : null });
}
var zn = "_1sj0tk80", zt = "_1sj0tk82", Kn = "_1sj0tk85", Kt = "_1sj0tk81", Q = {}, Yn = function() {
  return typeof Promise == "function" && Promise.prototype && Promise.prototype.then;
}, ae = {}, F = {};
let Rt;
const $n = [
  0,
  // Not used
  26,
  44,
  70,
  100,
  134,
  172,
  196,
  242,
  292,
  346,
  404,
  466,
  532,
  581,
  655,
  733,
  815,
  901,
  991,
  1085,
  1156,
  1258,
  1364,
  1474,
  1588,
  1706,
  1828,
  1921,
  2051,
  2185,
  2323,
  2465,
  2611,
  2761,
  2876,
  3034,
  3196,
  3362,
  3532,
  3706
];
F.getSymbolSize = function(e) {
  if (!e) throw new Error('"version" cannot be null or undefined');
  if (e < 1 || e > 40) throw new Error('"version" should be in range from 1 to 40');
  return e * 4 + 17;
};
F.getSymbolTotalCodewords = function(e) {
  return $n[e];
};
F.getBCHDigit = function(t) {
  let e = 0;
  for (; t !== 0; )
    e++, t >>>= 1;
  return e;
};
F.setToSJISFunction = function(e) {
  if (typeof e != "function")
    throw new Error('"toSJISFunc" is not a valid function.');
  Rt = e;
};
F.isKanjiModeEnabled = function() {
  return typeof Rt < "u";
};
F.toSJIS = function(e) {
  return Rt(e);
};
var ft = {};
(function(t) {
  t.L = { bit: 1 }, t.M = { bit: 0 }, t.Q = { bit: 3 }, t.H = { bit: 2 };
  function e(n) {
    if (typeof n != "string")
      throw new Error("Param is not a string");
    switch (n.toLowerCase()) {
      case "l":
      case "low":
        return t.L;
      case "m":
      case "medium":
        return t.M;
      case "q":
      case "quartile":
        return t.Q;
      case "h":
      case "high":
        return t.H;
      default:
        throw new Error("Unknown EC Level: " + n);
    }
  }
  t.isValid = function(r) {
    return r && typeof r.bit < "u" && r.bit >= 0 && r.bit < 4;
  }, t.from = function(r, o) {
    if (t.isValid(r))
      return r;
    try {
      return e(r);
    } catch {
      return o;
    }
  };
})(ft);
function ce() {
  this.buffer = [], this.length = 0;
}
ce.prototype = {
  get: function(t) {
    const e = Math.floor(t / 8);
    return (this.buffer[e] >>> 7 - t % 8 & 1) === 1;
  },
  put: function(t, e) {
    for (let n = 0; n < e; n++)
      this.putBit((t >>> e - n - 1 & 1) === 1);
  },
  getLengthInBits: function() {
    return this.length;
  },
  putBit: function(t) {
    const e = Math.floor(this.length / 8);
    this.buffer.length <= e && this.buffer.push(0), t && (this.buffer[e] |= 128 >>> this.length % 8), this.length++;
  }
};
var jn = ce;
function G(t) {
  if (!t || t < 1)
    throw new Error("BitMatrix size must be defined and greater than 0");
  this.size = t, this.data = new Uint8Array(t * t), this.reservedBit = new Uint8Array(t * t);
}
G.prototype.set = function(t, e, n, r) {
  const o = t * this.size + e;
  this.data[o] = n, r && (this.reservedBit[o] = !0);
};
G.prototype.get = function(t, e) {
  return this.data[t * this.size + e];
};
G.prototype.xor = function(t, e, n) {
  this.data[t * this.size + e] ^= n;
};
G.prototype.isReserved = function(t, e) {
  return this.reservedBit[t * this.size + e];
};
var Jn = G, se = {};
(function(t) {
  const e = F.getSymbolSize;
  t.getRowColCoords = function(r) {
    if (r === 1) return [];
    const o = Math.floor(r / 7) + 2, i = e(r), a = i === 145 ? 26 : Math.ceil((i - 13) / (2 * o - 2)) * 2, c = [i - 7];
    for (let s = 1; s < o - 1; s++)
      c[s] = c[s - 1] - a;
    return c.push(6), c.reverse();
  }, t.getPositions = function(r) {
    const o = [], i = t.getRowColCoords(r), a = i.length;
    for (let c = 0; c < a; c++)
      for (let s = 0; s < a; s++)
        c === 0 && s === 0 || // top-left
        c === 0 && s === a - 1 || // bottom-left
        c === a - 1 && s === 0 || o.push([i[c], i[s]]);
    return o;
  };
})(se);
var le = {};
const Wn = F.getSymbolSize, Yt = 7;
le.getPositions = function(e) {
  const n = Wn(e);
  return [
    // top-left
    [0, 0],
    // top-right
    [n - Yt, 0],
    // bottom-left
    [0, n - Yt]
  ];
};
var ue = {};
(function(t) {
  t.Patterns = {
    PATTERN000: 0,
    PATTERN001: 1,
    PATTERN010: 2,
    PATTERN011: 3,
    PATTERN100: 4,
    PATTERN101: 5,
    PATTERN110: 6,
    PATTERN111: 7
  };
  const e = {
    N1: 3,
    N2: 3,
    N3: 40,
    N4: 10
  };
  t.isValid = function(o) {
    return o != null && o !== "" && !isNaN(o) && o >= 0 && o <= 7;
  }, t.from = function(o) {
    return t.isValid(o) ? parseInt(o, 10) : void 0;
  }, t.getPenaltyN1 = function(o) {
    const i = o.size;
    let a = 0, c = 0, s = 0, f = null, l = null;
    for (let E = 0; E < i; E++) {
      c = s = 0, f = l = null;
      for (let w = 0; w < i; w++) {
        let d = o.get(E, w);
        d === f ? c++ : (c >= 5 && (a += e.N1 + (c - 5)), f = d, c = 1), d = o.get(w, E), d === l ? s++ : (s >= 5 && (a += e.N1 + (s - 5)), l = d, s = 1);
      }
      c >= 5 && (a += e.N1 + (c - 5)), s >= 5 && (a += e.N1 + (s - 5));
    }
    return a;
  }, t.getPenaltyN2 = function(o) {
    const i = o.size;
    let a = 0;
    for (let c = 0; c < i - 1; c++)
      for (let s = 0; s < i - 1; s++) {
        const f = o.get(c, s) + o.get(c, s + 1) + o.get(c + 1, s) + o.get(c + 1, s + 1);
        (f === 4 || f === 0) && a++;
      }
    return a * e.N2;
  }, t.getPenaltyN3 = function(o) {
    const i = o.size;
    let a = 0, c = 0, s = 0;
    for (let f = 0; f < i; f++) {
      c = s = 0;
      for (let l = 0; l < i; l++)
        c = c << 1 & 2047 | o.get(f, l), l >= 10 && (c === 1488 || c === 93) && a++, s = s << 1 & 2047 | o.get(l, f), l >= 10 && (s === 1488 || s === 93) && a++;
    }
    return a * e.N3;
  }, t.getPenaltyN4 = function(o) {
    let i = 0;
    const a = o.data.length;
    for (let s = 0; s < a; s++) i += o.data[s];
    return Math.abs(Math.ceil(i * 100 / a / 5) - 10) * e.N4;
  };
  function n(r, o, i) {
    switch (r) {
      case t.Patterns.PATTERN000:
        return (o + i) % 2 === 0;
      case t.Patterns.PATTERN001:
        return o % 2 === 0;
      case t.Patterns.PATTERN010:
        return i % 3 === 0;
      case t.Patterns.PATTERN011:
        return (o + i) % 3 === 0;
      case t.Patterns.PATTERN100:
        return (Math.floor(o / 2) + Math.floor(i / 3)) % 2 === 0;
      case t.Patterns.PATTERN101:
        return o * i % 2 + o * i % 3 === 0;
      case t.Patterns.PATTERN110:
        return (o * i % 2 + o * i % 3) % 2 === 0;
      case t.Patterns.PATTERN111:
        return (o * i % 3 + (o + i) % 2) % 2 === 0;
      default:
        throw new Error("bad maskPattern:" + r);
    }
  }
  t.applyMask = function(o, i) {
    const a = i.size;
    for (let c = 0; c < a; c++)
      for (let s = 0; s < a; s++)
        i.isReserved(s, c) || i.xor(s, c, n(o, s, c));
  }, t.getBestMask = function(o, i) {
    const a = Object.keys(t.Patterns).length;
    let c = 0, s = 1 / 0;
    for (let f = 0; f < a; f++) {
      i(f), t.applyMask(f, o);
      const l = t.getPenaltyN1(o) + t.getPenaltyN2(o) + t.getPenaltyN3(o) + t.getPenaltyN4(o);
      t.applyMask(f, o), l < s && (s = l, c = f);
    }
    return c;
  };
})(ue);
var dt = {};
const V = ft, rt = [
  // L  M  Q  H
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  2,
  2,
  1,
  2,
  2,
  4,
  1,
  2,
  4,
  4,
  2,
  4,
  4,
  4,
  2,
  4,
  6,
  5,
  2,
  4,
  6,
  6,
  2,
  5,
  8,
  8,
  4,
  5,
  8,
  8,
  4,
  5,
  8,
  11,
  4,
  8,
  10,
  11,
  4,
  9,
  12,
  16,
  4,
  9,
  16,
  16,
  6,
  10,
  12,
  18,
  6,
  10,
  17,
  16,
  6,
  11,
  16,
  19,
  6,
  13,
  18,
  21,
  7,
  14,
  21,
  25,
  8,
  16,
  20,
  25,
  8,
  17,
  23,
  25,
  9,
  17,
  23,
  34,
  9,
  18,
  25,
  30,
  10,
  20,
  27,
  32,
  12,
  21,
  29,
  35,
  12,
  23,
  34,
  37,
  12,
  25,
  34,
  40,
  13,
  26,
  35,
  42,
  14,
  28,
  38,
  45,
  15,
  29,
  40,
  48,
  16,
  31,
  43,
  51,
  17,
  33,
  45,
  54,
  18,
  35,
  48,
  57,
  19,
  37,
  51,
  60,
  19,
  38,
  53,
  63,
  20,
  40,
  56,
  66,
  21,
  43,
  59,
  70,
  22,
  45,
  62,
  74,
  24,
  47,
  65,
  77,
  25,
  49,
  68,
  81
], ot = [
  // L  M  Q  H
  7,
  10,
  13,
  17,
  10,
  16,
  22,
  28,
  15,
  26,
  36,
  44,
  20,
  36,
  52,
  64,
  26,
  48,
  72,
  88,
  36,
  64,
  96,
  112,
  40,
  72,
  108,
  130,
  48,
  88,
  132,
  156,
  60,
  110,
  160,
  192,
  72,
  130,
  192,
  224,
  80,
  150,
  224,
  264,
  96,
  176,
  260,
  308,
  104,
  198,
  288,
  352,
  120,
  216,
  320,
  384,
  132,
  240,
  360,
  432,
  144,
  280,
  408,
  480,
  168,
  308,
  448,
  532,
  180,
  338,
  504,
  588,
  196,
  364,
  546,
  650,
  224,
  416,
  600,
  700,
  224,
  442,
  644,
  750,
  252,
  476,
  690,
  816,
  270,
  504,
  750,
  900,
  300,
  560,
  810,
  960,
  312,
  588,
  870,
  1050,
  336,
  644,
  952,
  1110,
  360,
  700,
  1020,
  1200,
  390,
  728,
  1050,
  1260,
  420,
  784,
  1140,
  1350,
  450,
  812,
  1200,
  1440,
  480,
  868,
  1290,
  1530,
  510,
  924,
  1350,
  1620,
  540,
  980,
  1440,
  1710,
  570,
  1036,
  1530,
  1800,
  570,
  1064,
  1590,
  1890,
  600,
  1120,
  1680,
  1980,
  630,
  1204,
  1770,
  2100,
  660,
  1260,
  1860,
  2220,
  720,
  1316,
  1950,
  2310,
  750,
  1372,
  2040,
  2430
];
dt.getBlocksCount = function(e, n) {
  switch (n) {
    case V.L:
      return rt[(e - 1) * 4 + 0];
    case V.M:
      return rt[(e - 1) * 4 + 1];
    case V.Q:
      return rt[(e - 1) * 4 + 2];
    case V.H:
      return rt[(e - 1) * 4 + 3];
    default:
      return;
  }
};
dt.getTotalCodewordsCount = function(e, n) {
  switch (n) {
    case V.L:
      return ot[(e - 1) * 4 + 0];
    case V.M:
      return ot[(e - 1) * 4 + 1];
    case V.Q:
      return ot[(e - 1) * 4 + 2];
    case V.H:
      return ot[(e - 1) * 4 + 3];
    default:
      return;
  }
};
var fe = {}, ht = {};
const j = new Uint8Array(512), ct = new Uint8Array(256);
(function() {
  let e = 1;
  for (let n = 0; n < 255; n++)
    j[n] = e, ct[e] = n, e <<= 1, e & 256 && (e ^= 285);
  for (let n = 255; n < 512; n++)
    j[n] = j[n - 255];
})();
ht.log = function(e) {
  if (e < 1) throw new Error("log(" + e + ")");
  return ct[e];
};
ht.exp = function(e) {
  return j[e];
};
ht.mul = function(e, n) {
  return e === 0 || n === 0 ? 0 : j[ct[e] + ct[n]];
};
(function(t) {
  const e = ht;
  t.mul = function(r, o) {
    const i = new Uint8Array(r.length + o.length - 1);
    for (let a = 0; a < r.length; a++)
      for (let c = 0; c < o.length; c++)
        i[a + c] ^= e.mul(r[a], o[c]);
    return i;
  }, t.mod = function(r, o) {
    let i = new Uint8Array(r);
    for (; i.length - o.length >= 0; ) {
      const a = i[0];
      for (let s = 0; s < o.length; s++)
        i[s] ^= e.mul(o[s], a);
      let c = 0;
      for (; c < i.length && i[c] === 0; ) c++;
      i = i.slice(c);
    }
    return i;
  }, t.generateECPolynomial = function(r) {
    let o = new Uint8Array([1]);
    for (let i = 0; i < r; i++)
      o = t.mul(o, new Uint8Array([1, e.exp(i)]));
    return o;
  };
})(fe);
const de = fe;
function kt(t) {
  this.genPoly = void 0, this.degree = t, this.degree && this.initialize(this.degree);
}
kt.prototype.initialize = function(e) {
  this.degree = e, this.genPoly = de.generateECPolynomial(this.degree);
};
kt.prototype.encode = function(e) {
  if (!this.genPoly)
    throw new Error("Encoder not initialized");
  const n = new Uint8Array(e.length + this.degree);
  n.set(e);
  const r = de.mod(n, this.genPoly), o = this.degree - r.length;
  if (o > 0) {
    const i = new Uint8Array(this.degree);
    return i.set(r, o), i;
  }
  return r;
};
var qn = kt, he = {}, H = {}, Lt = {};
Lt.isValid = function(e) {
  return !isNaN(e) && e >= 1 && e <= 40;
};
var U = {};
const ge = "[0-9]+", Qn = "[A-Z $%*+\\-./:]+";
let W = "(?:[u3000-u303F]|[u3040-u309F]|[u30A0-u30FF]|[uFF00-uFFEF]|[u4E00-u9FAF]|[u2605-u2606]|[u2190-u2195]|u203B|[u2010u2015u2018u2019u2025u2026u201Cu201Du2225u2260]|[u0391-u0451]|[u00A7u00A8u00B1u00B4u00D7u00F7])+";
W = W.replace(/u/g, "\\u");
const Gn = "(?:(?![A-Z0-9 $%*+\\-./:]|" + W + `)(?:.|[\r
]))+`;
U.KANJI = new RegExp(W, "g");
U.BYTE_KANJI = new RegExp("[^A-Z0-9 $%*+\\-./:]+", "g");
U.BYTE = new RegExp(Gn, "g");
U.NUMERIC = new RegExp(ge, "g");
U.ALPHANUMERIC = new RegExp(Qn, "g");
const Xn = new RegExp("^" + W + "$"), tr = new RegExp("^" + ge + "$"), er = new RegExp("^[A-Z0-9 $%*+\\-./:]+$");
U.testKanji = function(e) {
  return Xn.test(e);
};
U.testNumeric = function(e) {
  return tr.test(e);
};
U.testAlphanumeric = function(e) {
  return er.test(e);
};
(function(t) {
  const e = Lt, n = U;
  t.NUMERIC = {
    id: "Numeric",
    bit: 1,
    ccBits: [10, 12, 14]
  }, t.ALPHANUMERIC = {
    id: "Alphanumeric",
    bit: 2,
    ccBits: [9, 11, 13]
  }, t.BYTE = {
    id: "Byte",
    bit: 4,
    ccBits: [8, 16, 16]
  }, t.KANJI = {
    id: "Kanji",
    bit: 8,
    ccBits: [8, 10, 12]
  }, t.MIXED = {
    bit: -1
  }, t.getCharCountIndicator = function(i, a) {
    if (!i.ccBits) throw new Error("Invalid mode: " + i);
    if (!e.isValid(a))
      throw new Error("Invalid version: " + a);
    return a >= 1 && a < 10 ? i.ccBits[0] : a < 27 ? i.ccBits[1] : i.ccBits[2];
  }, t.getBestModeForData = function(i) {
    return n.testNumeric(i) ? t.NUMERIC : n.testAlphanumeric(i) ? t.ALPHANUMERIC : n.testKanji(i) ? t.KANJI : t.BYTE;
  }, t.toString = function(i) {
    if (i && i.id) return i.id;
    throw new Error("Invalid mode");
  }, t.isValid = function(i) {
    return i && i.bit && i.ccBits;
  };
  function r(o) {
    if (typeof o != "string")
      throw new Error("Param is not a string");
    switch (o.toLowerCase()) {
      case "numeric":
        return t.NUMERIC;
      case "alphanumeric":
        return t.ALPHANUMERIC;
      case "kanji":
        return t.KANJI;
      case "byte":
        return t.BYTE;
      default:
        throw new Error("Unknown mode: " + o);
    }
  }
  t.from = function(i, a) {
    if (t.isValid(i))
      return i;
    try {
      return r(i);
    } catch {
      return a;
    }
  };
})(H);
(function(t) {
  const e = F, n = dt, r = ft, o = H, i = Lt, a = 7973, c = e.getBCHDigit(a);
  function s(w, d, g) {
    for (let u = 1; u <= 40; u++)
      if (d <= t.getCapacity(u, g, w))
        return u;
  }
  function f(w, d) {
    return o.getCharCountIndicator(w, d) + 4;
  }
  function l(w, d) {
    let g = 0;
    return w.forEach(function(u) {
      const y = f(u.mode, d);
      g += y + u.getBitsLength();
    }), g;
  }
  function E(w, d) {
    for (let g = 1; g <= 40; g++)
      if (l(w, g) <= t.getCapacity(g, d, o.MIXED))
        return g;
  }
  t.from = function(d, g) {
    return i.isValid(d) ? parseInt(d, 10) : g;
  }, t.getCapacity = function(d, g, u) {
    if (!i.isValid(d))
      throw new Error("Invalid QR Code version");
    typeof u > "u" && (u = o.BYTE);
    const y = e.getSymbolTotalCodewords(d), h = n.getTotalCodewordsCount(d, g), p = (y - h) * 8;
    if (u === o.MIXED) return p;
    const v = p - f(u, d);
    switch (u) {
      case o.NUMERIC:
        return Math.floor(v / 10 * 3);
      case o.ALPHANUMERIC:
        return Math.floor(v / 11 * 2);
      case o.KANJI:
        return Math.floor(v / 13);
      case o.BYTE:
      default:
        return Math.floor(v / 8);
    }
  }, t.getBestVersionForData = function(d, g) {
    let u;
    const y = r.from(g, r.M);
    if (Array.isArray(d)) {
      if (d.length > 1)
        return E(d, y);
      if (d.length === 0)
        return 1;
      u = d[0];
    } else
      u = d;
    return s(u.mode, u.getLength(), y);
  }, t.getEncodedBits = function(d) {
    if (!i.isValid(d) || d < 7)
      throw new Error("Invalid QR Code version");
    let g = d << 12;
    for (; e.getBCHDigit(g) - c >= 0; )
      g ^= a << e.getBCHDigit(g) - c;
    return d << 12 | g;
  };
})(he);
var me = {};
const At = F, ve = 1335, nr = 21522, $t = At.getBCHDigit(ve);
me.getEncodedBits = function(e, n) {
  const r = e.bit << 3 | n;
  let o = r << 10;
  for (; At.getBCHDigit(o) - $t >= 0; )
    o ^= ve << At.getBCHDigit(o) - $t;
  return (r << 10 | o) ^ nr;
};
var we = {};
const rr = H;
function z(t) {
  this.mode = rr.NUMERIC, this.data = t.toString();
}
z.getBitsLength = function(e) {
  return 10 * Math.floor(e / 3) + (e % 3 ? e % 3 * 3 + 1 : 0);
};
z.prototype.getLength = function() {
  return this.data.length;
};
z.prototype.getBitsLength = function() {
  return z.getBitsLength(this.data.length);
};
z.prototype.write = function(e) {
  let n, r, o;
  for (n = 0; n + 3 <= this.data.length; n += 3)
    r = this.data.substr(n, 3), o = parseInt(r, 10), e.put(o, 10);
  const i = this.data.length - n;
  i > 0 && (r = this.data.substr(n), o = parseInt(r, 10), e.put(o, i * 3 + 1));
};
var or = z;
const ir = H, yt = [
  "0",
  "1",
  "2",
  "3",
  "4",
  "5",
  "6",
  "7",
  "8",
  "9",
  "A",
  "B",
  "C",
  "D",
  "E",
  "F",
  "G",
  "H",
  "I",
  "J",
  "K",
  "L",
  "M",
  "N",
  "O",
  "P",
  "Q",
  "R",
  "S",
  "T",
  "U",
  "V",
  "W",
  "X",
  "Y",
  "Z",
  " ",
  "$",
  "%",
  "*",
  "+",
  "-",
  ".",
  "/",
  ":"
];
function K(t) {
  this.mode = ir.ALPHANUMERIC, this.data = t;
}
K.getBitsLength = function(e) {
  return 11 * Math.floor(e / 2) + 6 * (e % 2);
};
K.prototype.getLength = function() {
  return this.data.length;
};
K.prototype.getBitsLength = function() {
  return K.getBitsLength(this.data.length);
};
K.prototype.write = function(e) {
  let n;
  for (n = 0; n + 2 <= this.data.length; n += 2) {
    let r = yt.indexOf(this.data[n]) * 45;
    r += yt.indexOf(this.data[n + 1]), e.put(r, 11);
  }
  this.data.length % 2 && e.put(yt.indexOf(this.data[n]), 6);
};
var ar = K, cr = function(e) {
  for (var n = [], r = e.length, o = 0; o < r; o++) {
    var i = e.charCodeAt(o);
    if (i >= 55296 && i <= 56319 && r > o + 1) {
      var a = e.charCodeAt(o + 1);
      a >= 56320 && a <= 57343 && (i = (i - 55296) * 1024 + a - 56320 + 65536, o += 1);
    }
    if (i < 128) {
      n.push(i);
      continue;
    }
    if (i < 2048) {
      n.push(i >> 6 | 192), n.push(i & 63 | 128);
      continue;
    }
    if (i < 55296 || i >= 57344 && i < 65536) {
      n.push(i >> 12 | 224), n.push(i >> 6 & 63 | 128), n.push(i & 63 | 128);
      continue;
    }
    if (i >= 65536 && i <= 1114111) {
      n.push(i >> 18 | 240), n.push(i >> 12 & 63 | 128), n.push(i >> 6 & 63 | 128), n.push(i & 63 | 128);
      continue;
    }
    n.push(239, 191, 189);
  }
  return new Uint8Array(n).buffer;
};
const sr = cr, lr = H;
function Y(t) {
  this.mode = lr.BYTE, typeof t == "string" && (t = sr(t)), this.data = new Uint8Array(t);
}
Y.getBitsLength = function(e) {
  return e * 8;
};
Y.prototype.getLength = function() {
  return this.data.length;
};
Y.prototype.getBitsLength = function() {
  return Y.getBitsLength(this.data.length);
};
Y.prototype.write = function(t) {
  for (let e = 0, n = this.data.length; e < n; e++)
    t.put(this.data[e], 8);
};
var ur = Y;
const fr = H, dr = F;
function $(t) {
  this.mode = fr.KANJI, this.data = t;
}
$.getBitsLength = function(e) {
  return e * 13;
};
$.prototype.getLength = function() {
  return this.data.length;
};
$.prototype.getBitsLength = function() {
  return $.getBitsLength(this.data.length);
};
$.prototype.write = function(t) {
  let e;
  for (e = 0; e < this.data.length; e++) {
    let n = dr.toSJIS(this.data[e]);
    if (n >= 33088 && n <= 40956)
      n -= 33088;
    else if (n >= 57408 && n <= 60351)
      n -= 49472;
    else
      throw new Error(
        "Invalid SJIS character: " + this.data[e] + `
Make sure your charset is UTF-8`
      );
    n = (n >>> 8 & 255) * 192 + (n & 255), t.put(n, 13);
  }
};
var hr = $, pe = { exports: {} };
(function(t) {
  var e = {
    single_source_shortest_paths: function(n, r, o) {
      var i = {}, a = {};
      a[r] = 0;
      var c = e.PriorityQueue.make();
      c.push(r, 0);
      for (var s, f, l, E, w, d, g, u, y; !c.empty(); ) {
        s = c.pop(), f = s.value, E = s.cost, w = n[f] || {};
        for (l in w)
          w.hasOwnProperty(l) && (d = w[l], g = E + d, u = a[l], y = typeof a[l] > "u", (y || u > g) && (a[l] = g, c.push(l, g), i[l] = f));
      }
      if (typeof o < "u" && typeof a[o] > "u") {
        var h = ["Could not find a path from ", r, " to ", o, "."].join("");
        throw new Error(h);
      }
      return i;
    },
    extract_shortest_path_from_predecessor_list: function(n, r) {
      for (var o = [], i = r; i; )
        o.push(i), n[i], i = n[i];
      return o.reverse(), o;
    },
    find_path: function(n, r, o) {
      var i = e.single_source_shortest_paths(n, r, o);
      return e.extract_shortest_path_from_predecessor_list(
        i,
        o
      );
    },
    /**
     * A very naive priority queue implementation.
     */
    PriorityQueue: {
      make: function(n) {
        var r = e.PriorityQueue, o = {}, i;
        n = n || {};
        for (i in r)
          r.hasOwnProperty(i) && (o[i] = r[i]);
        return o.queue = [], o.sorter = n.sorter || r.default_sorter, o;
      },
      default_sorter: function(n, r) {
        return n.cost - r.cost;
      },
      /**
       * Add a new item to the queue and ensure the highest priority element
       * is at the front of the queue.
       */
      push: function(n, r) {
        var o = { value: n, cost: r };
        this.queue.push(o), this.queue.sort(this.sorter);
      },
      /**
       * Return the highest priority element in the queue.
       */
      pop: function() {
        return this.queue.shift();
      },
      empty: function() {
        return this.queue.length === 0;
      }
    }
  };
  t.exports = e;
})(pe);
var gr = pe.exports;
(function(t) {
  const e = H, n = or, r = ar, o = ur, i = hr, a = U, c = F, s = gr;
  function f(h) {
    return unescape(encodeURIComponent(h)).length;
  }
  function l(h, p, v) {
    const m = [];
    let S;
    for (; (S = h.exec(v)) !== null; )
      m.push({
        data: S[0],
        index: S.index,
        mode: p,
        length: S[0].length
      });
    return m;
  }
  function E(h) {
    const p = l(a.NUMERIC, e.NUMERIC, h), v = l(a.ALPHANUMERIC, e.ALPHANUMERIC, h);
    let m, S;
    return c.isKanjiModeEnabled() ? (m = l(a.BYTE, e.BYTE, h), S = l(a.KANJI, e.KANJI, h)) : (m = l(a.BYTE_KANJI, e.BYTE, h), S = []), p.concat(v, m, S).sort(function(b, A) {
      return b.index - A.index;
    }).map(function(b) {
      return {
        data: b.data,
        mode: b.mode,
        length: b.length
      };
    });
  }
  function w(h, p) {
    switch (p) {
      case e.NUMERIC:
        return n.getBitsLength(h);
      case e.ALPHANUMERIC:
        return r.getBitsLength(h);
      case e.KANJI:
        return i.getBitsLength(h);
      case e.BYTE:
        return o.getBitsLength(h);
    }
  }
  function d(h) {
    return h.reduce(function(p, v) {
      const m = p.length - 1 >= 0 ? p[p.length - 1] : null;
      return m && m.mode === v.mode ? (p[p.length - 1].data += v.data, p) : (p.push(v), p);
    }, []);
  }
  function g(h) {
    const p = [];
    for (let v = 0; v < h.length; v++) {
      const m = h[v];
      switch (m.mode) {
        case e.NUMERIC:
          p.push([
            m,
            { data: m.data, mode: e.ALPHANUMERIC, length: m.length },
            { data: m.data, mode: e.BYTE, length: m.length }
          ]);
          break;
        case e.ALPHANUMERIC:
          p.push([
            m,
            { data: m.data, mode: e.BYTE, length: m.length }
          ]);
          break;
        case e.KANJI:
          p.push([
            m,
            { data: m.data, mode: e.BYTE, length: f(m.data) }
          ]);
          break;
        case e.BYTE:
          p.push([
            { data: m.data, mode: e.BYTE, length: f(m.data) }
          ]);
      }
    }
    return p;
  }
  function u(h, p) {
    const v = {}, m = { start: {} };
    let S = ["start"];
    for (let B = 0; B < h.length; B++) {
      const b = h[B], A = [];
      for (let M = 0; M < b.length; M++) {
        const N = b[M], k = "" + B + M;
        A.push(k), v[k] = { node: N, lastCount: 0 }, m[k] = {};
        for (let L = 0; L < S.length; L++) {
          const _ = S[L];
          v[_] && v[_].node.mode === N.mode ? (m[_][k] = w(v[_].lastCount + N.length, N.mode) - w(v[_].lastCount, N.mode), v[_].lastCount += N.length) : (v[_] && (v[_].lastCount = N.length), m[_][k] = w(N.length, N.mode) + 4 + e.getCharCountIndicator(N.mode, p));
        }
      }
      S = A;
    }
    for (let B = 0; B < S.length; B++)
      m[S[B]].end = 0;
    return { map: m, table: v };
  }
  function y(h, p) {
    let v;
    const m = e.getBestModeForData(h);
    if (v = e.from(p, m), v !== e.BYTE && v.bit < m.bit)
      throw new Error('"' + h + '" cannot be encoded with mode ' + e.toString(v) + `.
 Suggested mode is: ` + e.toString(m));
    switch (v === e.KANJI && !c.isKanjiModeEnabled() && (v = e.BYTE), v) {
      case e.NUMERIC:
        return new n(h);
      case e.ALPHANUMERIC:
        return new r(h);
      case e.KANJI:
        return new i(h);
      case e.BYTE:
        return new o(h);
    }
  }
  t.fromArray = function(p) {
    return p.reduce(function(v, m) {
      return typeof m == "string" ? v.push(y(m, null)) : m.data && v.push(y(m.data, m.mode)), v;
    }, []);
  }, t.fromString = function(p, v) {
    const m = E(p, c.isKanjiModeEnabled()), S = g(m), B = u(S, v), b = s.find_path(B.map, "start", "end"), A = [];
    for (let M = 1; M < b.length - 1; M++)
      A.push(B.table[b[M]].node);
    return t.fromArray(d(A));
  }, t.rawSplit = function(p) {
    return t.fromArray(
      E(p, c.isKanjiModeEnabled())
    );
  };
})(we);
const gt = F, Ct = ft, mr = jn, vr = Jn, wr = se, pr = le, Mt = ue, Nt = dt, yr = qn, st = he, Cr = me, Er = H, Et = we;
function Sr(t, e) {
  const n = t.size, r = pr.getPositions(e);
  for (let o = 0; o < r.length; o++) {
    const i = r[o][0], a = r[o][1];
    for (let c = -1; c <= 7; c++)
      if (!(i + c <= -1 || n <= i + c))
        for (let s = -1; s <= 7; s++)
          a + s <= -1 || n <= a + s || (c >= 0 && c <= 6 && (s === 0 || s === 6) || s >= 0 && s <= 6 && (c === 0 || c === 6) || c >= 2 && c <= 4 && s >= 2 && s <= 4 ? t.set(i + c, a + s, !0, !0) : t.set(i + c, a + s, !1, !0));
  }
}
function br(t) {
  const e = t.size;
  for (let n = 8; n < e - 8; n++) {
    const r = n % 2 === 0;
    t.set(n, 6, r, !0), t.set(6, n, r, !0);
  }
}
function Br(t, e) {
  const n = wr.getPositions(e);
  for (let r = 0; r < n.length; r++) {
    const o = n[r][0], i = n[r][1];
    for (let a = -2; a <= 2; a++)
      for (let c = -2; c <= 2; c++)
        a === -2 || a === 2 || c === -2 || c === 2 || a === 0 && c === 0 ? t.set(o + a, i + c, !0, !0) : t.set(o + a, i + c, !1, !0);
  }
}
function Ir(t, e) {
  const n = t.size, r = st.getEncodedBits(e);
  let o, i, a;
  for (let c = 0; c < 18; c++)
    o = Math.floor(c / 3), i = c % 3 + n - 8 - 3, a = (r >> c & 1) === 1, t.set(o, i, a, !0), t.set(i, o, a, !0);
}
function St(t, e, n) {
  const r = t.size, o = Cr.getEncodedBits(e, n);
  let i, a;
  for (i = 0; i < 15; i++)
    a = (o >> i & 1) === 1, i < 6 ? t.set(i, 8, a, !0) : i < 8 ? t.set(i + 1, 8, a, !0) : t.set(r - 15 + i, 8, a, !0), i < 8 ? t.set(8, r - i - 1, a, !0) : i < 9 ? t.set(8, 15 - i - 1 + 1, a, !0) : t.set(8, 15 - i - 1, a, !0);
  t.set(r - 8, 8, 1, !0);
}
function Ar(t, e) {
  const n = t.size;
  let r = -1, o = n - 1, i = 7, a = 0;
  for (let c = n - 1; c > 0; c -= 2)
    for (c === 6 && c--; ; ) {
      for (let s = 0; s < 2; s++)
        if (!t.isReserved(o, c - s)) {
          let f = !1;
          a < e.length && (f = (e[a] >>> i & 1) === 1), t.set(o, c - s, f), i--, i === -1 && (a++, i = 7);
        }
      if (o += r, o < 0 || n <= o) {
        o -= r, r = -r;
        break;
      }
    }
}
function Mr(t, e, n) {
  const r = new mr();
  n.forEach(function(s) {
    r.put(s.mode.bit, 4), r.put(s.getLength(), Er.getCharCountIndicator(s.mode, t)), s.write(r);
  });
  const o = gt.getSymbolTotalCodewords(t), i = Nt.getTotalCodewordsCount(t, e), a = (o - i) * 8;
  for (r.getLengthInBits() + 4 <= a && r.put(0, 4); r.getLengthInBits() % 8 !== 0; )
    r.putBit(0);
  const c = (a - r.getLengthInBits()) / 8;
  for (let s = 0; s < c; s++)
    r.put(s % 2 ? 17 : 236, 8);
  return Nr(r, t, e);
}
function Nr(t, e, n) {
  const r = gt.getSymbolTotalCodewords(e), o = Nt.getTotalCodewordsCount(e, n), i = r - o, a = Nt.getBlocksCount(e, n), c = r % a, s = a - c, f = Math.floor(r / a), l = Math.floor(i / a), E = l + 1, w = f - l, d = new yr(w);
  let g = 0;
  const u = new Array(a), y = new Array(a);
  let h = 0;
  const p = new Uint8Array(t.buffer);
  for (let b = 0; b < a; b++) {
    const A = b < s ? l : E;
    u[b] = p.slice(g, g + A), y[b] = d.encode(u[b]), g += A, h = Math.max(h, A);
  }
  const v = new Uint8Array(r);
  let m = 0, S, B;
  for (S = 0; S < h; S++)
    for (B = 0; B < a; B++)
      S < u[B].length && (v[m++] = u[B][S]);
  for (S = 0; S < w; S++)
    for (B = 0; B < a; B++)
      v[m++] = y[B][S];
  return v;
}
function Pr(t, e, n, r) {
  let o;
  if (Array.isArray(t))
    o = Et.fromArray(t);
  else if (typeof t == "string") {
    let f = e;
    if (!f) {
      const l = Et.rawSplit(t);
      f = st.getBestVersionForData(l, n);
    }
    o = Et.fromString(t, f || 40);
  } else
    throw new Error("Invalid data");
  const i = st.getBestVersionForData(o, n);
  if (!i)
    throw new Error("The amount of data is too big to be stored in a QR Code");
  if (!e)
    e = i;
  else if (e < i)
    throw new Error(
      `
The chosen QR Code version cannot contain this amount of data.
Minimum version required to store current data is: ` + i + `.
`
    );
  const a = Mr(e, n, o), c = gt.getSymbolSize(e), s = new vr(c);
  return Sr(s, e), br(s), Br(s, e), St(s, n, 0), e >= 7 && Ir(s, e), Ar(s, a), isNaN(r) && (r = Mt.getBestMask(
    s,
    St.bind(null, s, n)
  )), Mt.applyMask(r, s), St(s, n, r), {
    modules: s,
    version: e,
    errorCorrectionLevel: n,
    maskPattern: r,
    segments: o
  };
}
ae.create = function(e, n) {
  if (typeof e > "u" || e === "")
    throw new Error("No input text");
  let r = Ct.M, o, i;
  return typeof n < "u" && (r = Ct.from(n.errorCorrectionLevel, Ct.M), o = st.from(n.version), i = Mt.from(n.maskPattern), n.toSJISFunc && gt.setToSJISFunction(n.toSJISFunc)), Pr(e, o, r, i);
};
var ye = {}, _t = {};
(function(t) {
  function e(n) {
    if (typeof n == "number" && (n = n.toString()), typeof n != "string")
      throw new Error("Color should be defined as hex string");
    let r = n.slice().replace("#", "").split("");
    if (r.length < 3 || r.length === 5 || r.length > 8)
      throw new Error("Invalid hex color: " + n);
    (r.length === 3 || r.length === 4) && (r = Array.prototype.concat.apply([], r.map(function(i) {
      return [i, i];
    }))), r.length === 6 && r.push("F", "F");
    const o = parseInt(r.join(""), 16);
    return {
      r: o >> 24 & 255,
      g: o >> 16 & 255,
      b: o >> 8 & 255,
      a: o & 255,
      hex: "#" + r.slice(0, 6).join("")
    };
  }
  t.getOptions = function(r) {
    r || (r = {}), r.color || (r.color = {});
    const o = typeof r.margin > "u" || r.margin === null || r.margin < 0 ? 4 : r.margin, i = r.width && r.width >= 21 ? r.width : void 0, a = r.scale || 4;
    return {
      width: i,
      scale: i ? 4 : a,
      margin: o,
      color: {
        dark: e(r.color.dark || "#000000ff"),
        light: e(r.color.light || "#ffffffff")
      },
      type: r.type,
      rendererOpts: r.rendererOpts || {}
    };
  }, t.getScale = function(r, o) {
    return o.width && o.width >= r + o.margin * 2 ? o.width / (r + o.margin * 2) : o.scale;
  }, t.getImageWidth = function(r, o) {
    const i = t.getScale(r, o);
    return Math.floor((r + o.margin * 2) * i);
  }, t.qrToImageData = function(r, o, i) {
    const a = o.modules.size, c = o.modules.data, s = t.getScale(a, i), f = Math.floor((a + i.margin * 2) * s), l = i.margin * s, E = [i.color.light, i.color.dark];
    for (let w = 0; w < f; w++)
      for (let d = 0; d < f; d++) {
        let g = (w * f + d) * 4, u = i.color.light;
        if (w >= l && d >= l && w < f - l && d < f - l) {
          const y = Math.floor((w - l) / s), h = Math.floor((d - l) / s);
          u = E[c[y * a + h] ? 1 : 0];
        }
        r[g++] = u.r, r[g++] = u.g, r[g++] = u.b, r[g] = u.a;
      }
  };
})(_t);
(function(t) {
  const e = _t;
  function n(o, i, a) {
    o.clearRect(0, 0, i.width, i.height), i.style || (i.style = {}), i.height = a, i.width = a, i.style.height = a + "px", i.style.width = a + "px";
  }
  function r() {
    try {
      return document.createElement("canvas");
    } catch {
      throw new Error("You need to specify a canvas element");
    }
  }
  t.render = function(i, a, c) {
    let s = c, f = a;
    typeof s > "u" && (!a || !a.getContext) && (s = a, a = void 0), a || (f = r()), s = e.getOptions(s);
    const l = e.getImageWidth(i.modules.size, s), E = f.getContext("2d"), w = E.createImageData(l, l);
    return e.qrToImageData(w.data, i, s), n(E, f, l), E.putImageData(w, 0, 0), f;
  }, t.renderToDataURL = function(i, a, c) {
    let s = c;
    typeof s > "u" && (!a || !a.getContext) && (s = a, a = void 0), s || (s = {});
    const f = t.render(i, a, s), l = s.type || "image/png", E = s.rendererOpts || {};
    return f.toDataURL(l, E.quality);
  };
})(ye);
var Ce = {};
const Tr = _t;
function jt(t, e) {
  const n = t.a / 255, r = e + '="' + t.hex + '"';
  return n < 1 ? r + " " + e + '-opacity="' + n.toFixed(2).slice(1) + '"' : r;
}
function bt(t, e, n) {
  let r = t + e;
  return typeof n < "u" && (r += " " + n), r;
}
function Rr(t, e, n) {
  let r = "", o = 0, i = !1, a = 0;
  for (let c = 0; c < t.length; c++) {
    const s = Math.floor(c % e), f = Math.floor(c / e);
    !s && !i && (i = !0), t[c] ? (a++, c > 0 && s > 0 && t[c - 1] || (r += i ? bt("M", s + n, 0.5 + f + n) : bt("m", o, 0), o = 0, i = !1), s + 1 < e && t[c + 1] || (r += bt("h", a), a = 0)) : o++;
  }
  return r;
}
Ce.render = function(e, n, r) {
  const o = Tr.getOptions(n), i = e.modules.size, a = e.modules.data, c = i + o.margin * 2, s = o.color.light.a ? "<path " + jt(o.color.light, "fill") + ' d="M0 0h' + c + "v" + c + 'H0z"/>' : "", f = "<path " + jt(o.color.dark, "stroke") + ' d="' + Rr(a, i, o.margin) + '"/>', l = 'viewBox="0 0 ' + c + " " + c + '"', w = '<svg xmlns="http://www.w3.org/2000/svg" ' + (o.width ? 'width="' + o.width + '" height="' + o.width + '" ' : "") + l + ' shape-rendering="crispEdges">' + s + f + `</svg>
`;
  return typeof r == "function" && r(null, w), w;
};
const kr = Yn, Pt = ae, Ee = ye, Lr = Ce;
function Ft(t, e, n, r, o) {
  const i = [].slice.call(arguments, 1), a = i.length, c = typeof i[a - 1] == "function";
  if (!c && !kr())
    throw new Error("Callback required as last argument");
  if (c) {
    if (a < 2)
      throw new Error("Too few arguments provided");
    a === 2 ? (o = n, n = e, e = r = void 0) : a === 3 && (e.getContext && typeof o > "u" ? (o = r, r = void 0) : (o = r, r = n, n = e, e = void 0));
  } else {
    if (a < 1)
      throw new Error("Too few arguments provided");
    return a === 1 ? (n = e, e = r = void 0) : a === 2 && !e.getContext && (r = n, n = e, e = void 0), new Promise(function(s, f) {
      try {
        const l = Pt.create(n, r);
        s(t(l, e, r));
      } catch (l) {
        f(l);
      }
    });
  }
  try {
    const s = Pt.create(n, r);
    o(null, t(s, e, r));
  } catch (s) {
    o(s);
  }
}
Q.create = Pt.create;
Q.toCanvas = Ft.bind(null, Ee.render);
Q.toDataURL = Ft.bind(null, Ee.renderToDataURL);
Q.toString = Ft.bind(null, function(t, e, n) {
  return Lr.render(t, n);
});
const _r = (t, e) => {
  const n = Array.prototype.slice.call(
    Q.create(t, { errorCorrectionLevel: e }).modules.data,
    0
  ), r = Math.sqrt(n.length);
  return n.reduce(
    (o, i, a) => (a % r === 0 ? o.push([i]) : o[o.length - 1].push(i)) && o,
    []
  );
};
function Fr({ ecl: t = "M", size: e = 200, uri: n }) {
  const o = e - parseInt("20", 10) * 2, i = Ae(() => {
    const a = [], c = _r(n, t), s = o / c.length;
    return [
      { x: 0, y: 0 },
      { x: 1, y: 0 },
      { x: 0, y: 1 }
    ].forEach(({ x: l, y: E }) => {
      const w = (c.length - 7) * s * l, d = (c.length - 7) * s * E;
      for (let g = 0; g < 3; g++)
        a.push(
          /* @__PURE__ */ C(
            "rect",
            {
              fill: g % 2 !== 0 ? "white" : "black",
              height: s * (7 - g * 2),
              width: s * (7 - g * 2),
              x: w + s * g,
              y: d + s * g
            },
            `${g}-${l}-${E}`
          )
        );
    }), c.forEach((l, E) => {
      l.forEach((w, d) => {
        c[E][d] && (E < 7 && d < 7 || E > c.length - 8 && d < 7 || E < 7 && d > c.length - 8 || a.push(
          /* @__PURE__ */ C(
            "rect",
            {
              fill: "black",
              height: s,
              width: s,
              x: E * s,
              y: d * s
            },
            `square-${E}-${d}`
          )
        ));
      });
    }), a;
  }, [t, o, n]);
  return /* @__PURE__ */ T("div", { className: Ve, children: [
    /* @__PURE__ */ C(
      "div",
      {
        className: Ue,
        style: {
          width: o
        }
      }
    ),
    /* @__PURE__ */ T("svg", { height: o, style: { all: "revert" }, width: o, children: [
      /* @__PURE__ */ C("title", { children: "QR Code" }),
      /* @__PURE__ */ C("rect", { fill: "transparent", height: o, width: o }),
      i
    ] })
  ] });
}
function Zr({
  open: t,
  onClose: e,
  url: n,
  isError: r,
  error: o
}) {
  return /* @__PURE__ */ C(xn, { open: t, titleId: "Sign in with Farcaster", onClose: e, children: /* @__PURE__ */ C("div", { className: "fc-authkit-qrcode-dialog", children: /* @__PURE__ */ C("div", { className: zn, children: r ? /* @__PURE__ */ T(J, { children: [
    /* @__PURE__ */ C("div", { className: Kt, children: "Error" }),
    /* @__PURE__ */ C("div", { className: zt, children: (o == null ? void 0 : o.message) ?? "Unknown error, please try again." })
  ] }) : /* @__PURE__ */ T(J, { children: [
    /* @__PURE__ */ C("div", { className: Kt, children: "Sign in with Farcaster" }),
    /* @__PURE__ */ C("div", { className: zt, children: "To sign in with Farcaster, scan the code below with your phone's camera." }),
    /* @__PURE__ */ C("div", { className: Kn, children: /* @__PURE__ */ C(Fr, { uri: n, size: 200 }) }),
    /* @__PURE__ */ C("div", { style: { display: "flex", justifyContent: "center" }, children: /* @__PURE__ */ T(
      Qt,
      {
        kind: "tertiary",
        style: {
          display: "flex",
          alignItems: "center",
          fontWeight: 500
        },
        onClick: () => {
          window.open(n, "_blank");
        },
        children: [
          /* @__PURE__ */ T(
            "svg",
            {
              xmlns: "http://www.w3.org/2000/svg",
              width: 12,
              height: 18,
              fill: "none",
              children: [
                /* @__PURE__ */ C("title", { children: "Sign in With Farcaster QR Code" }),
                /* @__PURE__ */ C(
                  "path",
                  {
                    fill: "#7C65C1",
                    fillRule: "evenodd",
                    d: "M0 3a3 3 0 0 1 3-3h6a3 3 0 0 1 3 3v12a3 3 0 0 1-3 3H3a3 3 0 0 1-3-3V3Zm4-1.5v.75c0 .414.336.75.75.75h2.5A.75.75 0 0 0 8 2.25V1.5h1A1.5 1.5 0 0 1 10.5 3v12A1.5 1.5 0 0 1 9 16.5H3A1.5 1.5 0 0 1 1.5 15V3A1.5 1.5 0 0 1 3 1.5h1Z",
                    clipRule: "evenodd"
                  }
                )
              ]
            }
          ),
          /* @__PURE__ */ C("span", { style: { marginLeft: 9 }, children: "I'm using my phone " })
        ]
      }
    ) })
  ] }) }) }) });
}
var Dr = "_8rz8bk0";
function Ur({
  debug: t,
  hideSignOut: e,
  onSignOut: n,
  ...r
}) {
  const { onSuccess: o, onStatusResponse: i, onError: a } = r, c = R(
    (L) => {
      o == null || o(L);
    },
    [o]
  ), s = R(
    (L) => {
      i == null || i(L);
    },
    [i]
  ), f = R(
    (L) => {
      a == null || a(L);
    },
    [a]
  ), l = R(() => {
    n == null || n();
  }, [n]), E = Fe({
    ...r,
    onSuccess: c,
    onStatusResponse: s,
    onError: f
  }), {
    signIn: w,
    signOut: d,
    connect: g,
    reconnect: u,
    isSuccess: y,
    isError: h,
    error: p,
    channelToken: v,
    url: m,
    data: S,
    validSignature: B
  } = E, b = R(() => {
    M(!1), d(), l();
  }, [d, l]), [A, M] = P(!1), N = R(() => {
    h && u(), M(!0), w(), m && It() && window.open(m, "_blank");
  }, [h, u, w, m]), k = y && B;
  return Z(() => {
    v || g();
  }, [v, g]), /* @__PURE__ */ T("div", { className: "fc-authkit-signin-button", children: [
    k ? /* @__PURE__ */ C(
      Qe,
      {
        userData: S,
        signOut: b,
        hideSignOut: !!e
      }
    ) : /* @__PURE__ */ T(J, { children: [
      /* @__PURE__ */ C(ze, { initializing: !m, onClick: N, label: "Sign in" }),
      m && /* @__PURE__ */ C(
        Zr,
        {
          open: A && !It(),
          onClose: () => M(!1),
          url: m,
          isError: h,
          error: p
        }
      )
    ] }),
    t && /* @__PURE__ */ C("div", { className: Dr, children: /* @__PURE__ */ C("pre", { children: JSON.stringify(E, null, 2) }) })
  ] });
}
function Vr() {
  const { isAuthenticated: t, profile: e } = q();
  return { isAuthenticated: t, profile: e };
}
function Yr() {
  return /* @__PURE__ */ T(Pe, { config: {
    rpcUrl: "https://mainnet.optimism.io"
  }, children: [
    /* @__PURE__ */ C("div", { style: { position: "fixed", top: 12, right: 12, zIndex: 1e3 }, children: /* @__PURE__ */ C(
      Ur,
      {
        nonce: "abcd1234",
        requestId: "some-unique-request-id",
        timeout: 2e4,
        onError: (e) => console.error("error callback:", e),
        onSuccess: (e) => console.log("success callback:", e),
        onStatusResponse: (e) => console.log("status callback:", e),
        onSignOut: () => console.log("sign out callback"),
        debug: !0
      }
    ) }),
    /* @__PURE__ */ C(Hr, {})
  ] });
}
function Hr() {
  const {
    isAuthenticated: t,
    profile: { fid: e, bio: n, displayName: r, custody: o }
  } = Vr();
  return /* @__PURE__ */ C("div", { style: { fontFamily: "sans-serif" }, children: t && /* @__PURE__ */ T("div", { children: [
    e && /* @__PURE__ */ T("p", { children: [
      /* @__PURE__ */ C("strong", { children: "FID:" }),
      " ",
      e
    ] }),
    r && /* @__PURE__ */ T("p", { children: [
      /* @__PURE__ */ C("strong", { children: "Display name:" }),
      " ",
      r
    ] }),
    n && /* @__PURE__ */ T("p", { children: [
      /* @__PURE__ */ C("strong", { children: "Bio:" }),
      " ",
      n
    ] }),
    o && /* @__PURE__ */ T("p", { children: [
      /* @__PURE__ */ C("strong", { children: "Custody address:" }),
      " ",
      o
    ] })
  ] }) });
}
function $r() {
  const { signInMessage: t } = q();
  return t;
}
export {
  Pe as AuthKitProvider,
  Yr as Demo,
  Fr as QRCode,
  Ur as SignInButton,
  lt as useAppClient,
  Te as useCreateChannel,
  Vr as useProfile,
  Fe as useSignIn,
  $r as useSignInMessage,
  Re as useVerifySignInMessage,
  Le as useWatchStatus
};
